ubg/src/pages/Room.js

```diff
@@ -1,126 +1,252 @@
+import firebase from 'firebase/app';
+import Paper from '@material-ui/core/Paper';
+import ChatIcon from '@material-ui/icons/Chat';
+import Chat from '../common/Chat';
+import NotFound from '../pages/NotFound';
+import GameRoom from '../common/GameRoom';
+import WaitingRoom from '../common/WaitingRoom';
 
 const useStyles = makeStyles((theme) => ({
   main: {
     height: '100vh',
   },
   video: {
-    background: 'black',
+    background: theme.palette.primary.main,
     textAlign: '-webkit-center',
   },
   gameContainer: {
     display: 'flex',
     flexDirection: 'column',
     position: 'relative',
   },
-  title: {
-    textAlign: 'center',
-    margin: '20px',
+  btn: {
+    margin: '5px',
   },
-  game: {
+  flexColumn: {
+    display: 'flex',
+    flexDirection: 'column',
+    height: '100%',
+  },
+  grow: {
     flexGrow: 1,
+  },
+  signInContainer: {
+    height: '100vh',
     display: 'flex',
-    justifyContent: 'center',
+    flexDirection: 'column',
     alignItems: 'center',
+    justifyContent: 'center',
   },
-  table: {
-    height: '300px',
-    width: '300px',
-    color: 'white',
-    background: 'black',
+  chatHeader: {
+    display: 'flex',
+    alignItems: 'center',
+    background: '#e0e0e0',
+    color: 'black',
+    borderRadius: '4px 4px 0 0',
   },
-  rules: {
-    position: 'absolute',
-    bottom: 0,
-    width: '100%',
-    textAlign: 'center',
+  sideMargin10px: {
+    margin: '0 10px',
   },
 }));
 
 /**
- * @return {ReactElement} Game room element
+ * @return {ReactElement} Room element
  */
 export default function Room() {
-  const [openRules, setOpenRules] = useState(false);
   const classes = useStyles();
-  const {uid, displayName, email} = useUser();
+  const user = useUser();
+  const auth = useAuth();
   const {roomId} = useParams();
   const room = useFirestore().collection('rooms').doc(roomId);
   const roomData = useFirestoreDocData(room);
   const usersCollection = room.collection('users');
   const usersData = useFirestoreCollectionData(usersCollection);
+  const prevUsersData = usePrevious(usersData);
+  const game = useFirestoreDocData(
+      useFirestore().collection('games').doc(roomData.gameId || ' '),
+  );
+  const [open, setOpen] = useState(false);
+  const [chatOpen, setChatOpen] = useState(false);
+  const [message, setMessage] = useState('');
+
+  /**
+   * Get previous value of variable
+   * @param {*} value
+   * @return {*} previous value
+   */
+  function usePrevious(value) {
+    const ref = useRef();
+    useEffect(() => {
+      ref.current = value;
+    });
+    return ref.current;
+  }
+
+  useEffect(() => {
+    if (prevUsersData) {
+      const usersJoined = usersData.filter((x) => !prevUsersData.includes(x));
+      const usersLeft = prevUsersData.filter((x) => !usersData.includes(x));
+      if (usersJoined.length > 0) {
+        setOpen(false);
+        setMessage(`${usersJoined[0].displayName} joined the room`);
+        setOpen(true);
```

Comment:
Why do you setOpen(false) then setOpen(true)?

Reply:
Because if the snackbar is open it gives the appearance of it closing and opening a new one, I'm going to add an if.